### DISCUSSION TITLE
Two-way communication with an Iphone

### DISCUSSION THREAD

**Discussion by PatLub85** [micropython/micropython#11377] [General]
Hi,
display an HTML page on an Iphone from an ESP Server via sockets, it works well and there are many tutorials.
My question is : can data be sent back from Iphone to ESP.
My goal is as follows : the ESP is a server in AP mode, lost at the bottom of my garden. When I connect the power to it, it loses the date and I would like to set the time from the Iphone.
Is it 'just' possible?

**Comment by bixb922** [comment]
You could embed a javascript in the page. This script can then get the date and send a http request with the date and time to the ESP32, preferably as Date.now() timestamp. Some timestamp arithmetic will be necessary on the ESP32.

AFAIK, ntptime cannot access internet when in AP mode.

Or you can activate a wifi hotspot on the iPhone and have the ESP32 connect to it in station mode and use ntptime to sync date and time.

**Comment by sophiedegran** [comment]
I use microdot on esp32c3 and rp2 as web server.
With it I can change date on these boards. I send the date on my client device (phone, tablet, computer).
I upload on these boards some html update, change frequency can be possible, and of course hard reset.
The most important is to deal a lot with javascript on html page before sending to server.
I use websocket include in microdot for that, except for files it's simply post.
Many examples on github.com/miguelgrinberg/microdot

**Comment by sophiedegran** [comment]
Ah I forgot one thing esp32 boards can be twice in AP mode and in STA mode and that's very usefull ☺

**Reply by PatLub85** [reply]
Thank you for your valuable information
At first I will try solution 2 of Bixb922 which seems the simplest.
However, I am interested in working with microdot. On the other hand, looking quickly at the documentation, I did not understand the interest of the first example (@app.route('/')), which is taken up by other Internet users. I understood that we had more to worry about sockets, but I did not quite understand the purpose of this framework

**Reply by bixb922** [reply]
Do you mean this example: https://github.com/miguelgrinberg/microdot/blob/main/examples/hello/hello.py

It is the simplest Webserver. If your microcontroller is at, say, 192.168.100.55, then http://192.168.100.55/ will respond with the page in the code. The argument of the @app.route decorator is the url after the address. Try it out!

**Reply by PatLub85** [reply]
In the examples I've seen, the server displays a web page to the client. On the other hand, I have not (or badly) seen the case where the ESP submits a form to which the client responds, for example in my case: month, day, hour, minutes.
Is it possible?

**Reply by karfas** [reply]
> the case where the ESP submits a form to which the client responds

You are mixing things up, I think. 

A **server** doesn't need to "submit forms"; form submission is usually done by a **client** (=browser).
However, you are free to send anything you like from the server side. This might include some JavaScript generated by the server and processed by the browser/client.
Or you serve a HTML form and answer (after pressing submit in the client) with a dynamic generated HTML page (which may or may not include JavaScript).

**Comment by sophiedegran** [comment]
I have a process which run microdot and other one which check sensors, he update little lists that are being graph in html. But he check also  determined lvl and if reached, active STA mode and send me an email. That's very usefull. I will test soon with lora (if no internet co. on server point).
